COVID19 RDT reference app docs, videos, training (#548)
medic/cht-docs: content/en/apps/examples/covid-rdt-reference-app.md

---

> title: "COVID-19 RDT Reference Application"
> linkTitle: "COVID-19 RDT"

Reviewer: Looking at the examples we have it may make sense to distinguish this as a "testing" workflow

```suggestion
title: "Reference Application using RDT for COVID-19 Testing"
linkTitle: "COVID-19 Testing"
```

weight: 
description: >
 A customizable CHT application with support for 3rd party RDT integration   
relatedContent: >
---

> Medic has worked with [FIND](https://www.finddx.org) to provide a reference application of rapid diagnostic tests (RDTs) for COVID-19 in the CHT.  By leveraging the expertise of third-party applications, like [Dimagi's RD-Toolkit](https://github.com/dimagi/rd-toolkit/), you can customize this application in your current or future deployment of the CHT.

Reviewer: Minor adjustment to language, and I think it's clearer to define the acronym without the `s` -- not sure if there is a standard way for that which we should follow though.

```suggestion
Medic has worked with [FIND](https://www.finddx.org) to provide a reference application using rapid diagnostic tests (RDT) for COVID-19 testing with the CHT.  By leveraging the expertise of third-party applications, like [Dimagi's RD-Toolkit](https://github.com/dimagi/rd-toolkit/), you can customize this application in your current or future deployment of the CHT.
```

Author: Yes! I went back and forth with the `s`, but happy to defer to your preference

Author: huh - couldn't accept your suggestion - exact same suggestion applied in 5d3f256

> You can find the code for this application in Medic's main [CHT Core repository on GitHub](https://github.com/medic/cht-core/tree/master/config/covid-19).

Reviewer: It would be helpful for us at Medic to be less possessive of the CHT ;)

```suggestion
You can find the code for this application in the [CHT Core repository on GitHub](https://github.com/medic/cht-core/tree/master/config/covid-19).
```

Author: \o/

## Problem Being Addressed

The original call for proposals best describes why Medic created this app: 

> FIND is working towards supporting countries in implementing an effective test-trace-isolate response using digital tools. To this end, FIND is looking to partner with leading digital solution providers to accelerate the development and deployment of a set of minimum functionalities for the collection of COVID RDT data and supporting incorporation of these functionalities into existing digital tools for use in low- and middle-income country settings.

## Solution Overview

This reference app provides a base layer of functionality that you can easily customize to meet the needs of your health program.  Specifically, this application has examples of:
 * Provisioning COVID-19 RDT tests 
 * Capturing COVID-19 RDT results
 * Storing all RDT data, including pictures, in the CHT

>  * Best practices of Health Facility, CHW, patient and RDT hierarchy in the CHT

Reviewer: I wasn't too clear on the original intention of this point. Here is my understanding, feel free to edit further.

```suggestion
 * Best practices for RDT workflows in the CHT
```

Author: yeah - this was me repurposing other "example" pages in the docs as templates.  Your change is most welcome!

### Images

27 high resolutions images taken from a demonstration CHW device are [available for download](CHT.COVID-19.RDT.Images.zip). These cover the entire usage of the reference app.  

### Videos

Provided here are two videos of the Provision and Capture forms in the COVID-19 application.

#### Provision 

This first video shows the left side of the workflow above to provision an RDT for a patient.  The CHW is shown finding Jessica Whitehouse's contact and choosing a new action of "RDT Covid-19 - Provision".  The CHW then does pre-test set up, checking Jessica's symptoms and location and confirming the test method and lot information. You can see the CHW launching the RD-Toolkit, reading the instructions and then, starting the session and timer.  With the RD-Toolkit sending back all the information to the CHT, a task "Capture Covid-19 RDT" can be seen in the CHT for the CHW to follow up on when the RD-Timer has completed. The last part of the video shows the completed provision report in the CHT:

<!-- 
  use raw HTML for now as hugo youtube short codes show 16:9 ration 
  box instead of mobile portrait ~9:16
-->

> <iframe id="ytplayer" type="text/html" width="290" height="610" src="https://www.youtube.com/embed/PGC0Mp6KtU4" frameborder="0"></iframe>

Author: Agreed! And I was trying to do the[ "cursor" in `scrcpy`](https://en.wikipedia.org/wiki/Scrcpy) - but of course you [can turn it on in android natively](https://www.skyneel.com/show-touches-in-android-phone).  I'll look into this!  As well,I'd be happy to do some edits to have "stages" and pause the video to show a interstitial white text on black background "CHW will now select the patient", "RDT test will not be provisioned including a timer and storing the session ID In the CHT" or some such thing.

Otherwise, if there's an in-house CHT demo video expert, I'd be happy to work with them!

#### Capture

> This second video shows the right side of the workflow above to capture RDT results for a patient.  The CHW is shown viewing the "Capture Covid-19 RDT" task after the 15 minute time from the RD-Timer has completed. After click the task, the CHW is brought to the CHT form which has the session information for the RD-Toolkit already loaded.  The test results are recorded and then returned to the CHT from the RD-Toolkit. The last part of the video shows the completed capture report in the CHT, including the prognosis and image of the RDT:

Author: Ah dude - I type `RD-Toolking` or `RD-Timer` or `RD-anthing-but-toolkit` so much! Indeed your fix is the correct one - thank you!

 * **API Calls:** Given that all raw forms captured are in JSON and given we know the data model well, you can easily do daily API calls to a CHT server instance and use some custom code (node, python etc) to gather and show stats on a daily basis. You can export to either JSON or CSV. See API docs [for reports]({{< ref "apps/reference/api#get-apiv2exportreports" >}}) as well as [monitoring metadata]({{< ref "apps/reference/api#get-apiv2monitoring" >}}). 
 * **[Postgres queries](https://github.com/medic/cht-couch2pg):** CHT ships with a utility to export all the data that the API has to a relational database, Postgres. You have all the raw data the API has, but can now use the power of joins and groupings to come up with totally customizable stats by day, week, month etc. Data can be synched daily or hourly from the CHT. 
 * **3rd party integrations:** We have used both [Klipfolio](https://www.klipfolio.com/) and, more recently, [Superset](https://superset.incubator.apache.org/) to create more user-friendly dashboards than any of the above options offer.  They talk to the Postgres server as the back end.

### Sample API call

> All calls to the API need to be authenticated. Be sure to follow our documentation on how to do this, but otherwise you need to use a login with `admin` permissions with this stucture: `http://LOGIN:PASSWORD@HOSTNAME`. If you had deployed this application without any customizations, you would have these forms available as seen by calling the [forms API]({{< ref "apps/reference/api#get-apiv1forms" >}}) (`/api/v1/forms`):

Author: Good points! Both fixed in 21e13b1

> All calls to the API need to be authenticated. Be sure to follow our documentation on how to do this, but otherwise you need to use a login with `admin` permissions with this stucture: `http://LOGIN:PASSWORD@HOSTNAME`. If you had deployed this application without any customizations, you would have these forms available as seen by calling the [forms API]({{< ref "apps/reference/api#get-apiv1forms" >}}) (`/api/v1/forms`):

Reviewer: Nice! I made a [comment in the commit](https://github.com/medic/cht-docs/commit/21e13b116607c3d6729294afa9d651fd05e742d1#r56798420).

* contact:health_center:create.xml
* contact:health_center:edit.xml
* contact:person:create.xml
* contact:person:edit.xml
* covid19_rdt_capture.xml
* covid19_rdt_provision.xml

The last two forms can be used to query [the reports API ]({{< ref "apps/reference/api#get-apiv2exportreports" >}}) to get RDT provions and captures.  To get all the capture reports into a file called `output.csv`, you would use this call:

```shell
curl "http://LOGIN:PASSWORD@HOSTNAME/api/v2/export/reports?filters[search]=&filters[forms][selected][0][code]=covid19_rdt_capture" > output.csv
```

### Base64 image extraction

To extract the binary image from the ASCII value in a report, you can view it in the CHT or you use a script to do so.  Here's an example of using Bash utilities (`sed`, `cut`, `tr` and `base64`) in Linux to get the image in the 1st row of data after the column headers into the file `image.jpg`. 

```shell
sed '2q;d' output.csv | cut -d',' -f15 | tr -d '"' | base64 -d > image.jpg
```

When retrieving [JSON](#capture-1), this value is found in `capture.android-app-outputs.rdt_session_bundle.rdt_session_result_bundle.rdt_session_result_extra_images.cropped` field.

> {{% alert title="Note" %}}The reports API always outputs in CSV, but the COVID-19 application uses [Base64 encoding](https://en.wikipedia.org/wiki/Base64) to store the images as text as noted above.  These may misbehave when opened them in a spreadsheet application like LibreOffice ("maximum number of characters per cell exceeded") or Excel (silently clipped to 32k chars) as they're thousands, if not hundreds of thousands, of characters long.  Be sure to programmatically process these into image files as needed. {{% /alert %}}

Reviewer: Do you have a script that you used internally already for this? It could be useful to post a command or snippet as an example rather than leaving it as an exercise from scratch.

Author: This is actually what I did!  The call to the reports API above results in a `outputs.csv` file.  The shell call in this section uses that same file to generate an image from one of the rows.

I don't think it's useful to bulk export the images from a CSV file (say in a while loop in Bash) as likely you'd be doing either an ETL into a 3rd party system, or just raw bulk import w/ no ETL and then on display show the base64 as an image, like we do with the CHT.

Maybe this wasn't clear though and I could add some extra explanation about this? Maybe you disagree and that a `while` script in Bash here is what's needed?

Reviewer: Right, not entirely _from scratch_, but from the command to extract a single image. It is likely not obvious from the extraction command what to change to get the other images out for a test, so to make it easier I think including the while loop would be helpful. If you think that there are too many aspects to take into account, then another option would be to mention which part to change to extract images from different lines and then if someone needs to do it programmatically they can figure it out from there.

Also, your comment about ETL is useful, and it could be worth mentioning in the text to give proper context.

Author: Oh - cool cool! thanks for the push-back on the push-back! I've tended to this in b71879f and b24dc48.  Lemme know what ya think!

Reviewer: Nice! Looks great!!

> ## Directory structure
> 
> Shown below is a filtered file structure of the COVID-19 app where you'll want to focus your customization efforts:

Reviewer: It wasn't clear to me what the "Directory structure" header was referring to, or what the filtered file structure criteria was. Here is a suggested adjustment:

```suggestion
## Customizing the reference app

These are the files in the COVID-19 app where you'll want to focus your customization efforts:
```

```
├── forms
│   ├── app
│   │   ├── covid19_rdt_capture.properties.json
│   │   ├── covid19_rdt_capture.xlsx
│   │   ├── covid19_rdt_capture.xml
│   │   ├── covid19_rdt_provision.properties.json
│   │   ├── covid19_rdt_provision.xlsx
│   │   └── covid19_rdt_provision.xml
├── tasks.js
```

The `forms/app/covid19_rdt_provision` and `forms/app/covid19_rdt_capture` forms (`xlsx`, `xml` and `properties.json`) represent the provision and capture portions of the forms.  The tasks that get created are defined in `tasks.js`.  Not shown are standard contact definitions in `forms/contact/*` as well as supporting configurations for icons and other CHT application settings.  

To read more about how these files all work together, see [app forms]({{< ref "apps/reference/forms/app" >}}), [contact forms]({{< ref "apps/reference/forms/contact" >}}), and [task]({{< ref "apps/reference/tasks" >}}) reference documentation

> ## JSON Reference 

Reviewer: It's nice to include the JSON of reports as reference. Given the "JSON Reference" header though, I was expecting JSON definition with descriptions, rather than a sample -- but that may just be me! Assuming it's not, then it may be nice to clarify the contents of this section in the header.

```suggestion
## Example form submission
```

## Solution Overview

This reference app provides a base layer of functionality that you can easily customize to meet the needs of your health program.  Specifically, this application has examples of:

 * Provisioning COVID-19 RDT tests 
 * Capturing COVID-19 RDT results
 * Storing all RDT data, including pictures, in the CHT
 * Best practices of Health Facility, CHW, patient and RDT hierarchy in the CHT
 
 Your instance of CHT needs to be on v3.13+ which has features developed to support this reference application, including:
 
  * **Display Base64 Image** - show ASCII encoded images inline
  * **Android App Launcher** - integrate with third party apps like the RD-Toolkit
  * **Parse Timestamp to Date** -  convert epoch time stamps (`1628945040308`) to your desired format (`Sun Mar 19 13:25:08`).
  
  For more information on these features, see the ["Related Content"](#related-content).

## Workflow

There are three main components to this application:
 * The CHT forms to provision and capture RDTs (Green)
 * Using a third-party app to use an RDT (Yellow)
 * Tasks to remind CHWs to complete an RDT that has been started (Blue) 

> {{< figure src="flow.png" link="flow.png" class="right col-12 col-lg-12" >}}

Reviewer: The diagram uses some notations differently than we do in other diagrams (eg blocks for _parts of_ a form, rather than individual forms). Let's chat directly about some changes that could simplify it.

Author: for sure! I knew that i was deviating the well trod-path, but thought it was worth it to denote the different apps, tasks and forms involved.

---

> title: "Reference Application using RDT for COVID-19 Testing"
> linkTitle: "COVID-19 Testing"
> weight: 
> description: >
>  A customizable CHT application with support for 3rd party RDT integration   

Reviewer: Revisiting to align with titles and descriptions in [index page](https://docs.communityhealthtoolkit.org/apps/examples/).

```suggestion
title: "COVID-19 Testing with Rapid Diagnostic Tests"
linkTitle: "COVID-19 Testing"
weight: 
description: >
 CHT example application that uses a third party app to capture the result of a Rapid Diagnostic Test.
```

relatedContent: >
  apps/reference/forms/app/#android-app-launcher
  apps/reference/forms/app/#cht-xform-widgets
  apps/reference/forms/app/#parse-timestamp-to-date
---

> Medic has worked with [FIND](https://www.finddx.org) to provide a reference application using rapid diagnostic tests (RDT) for COVID-19 testing with the CHT.  By leveraging the expertise of third-party applications, like [Dimagi's RD-Toolkit](https://github.com/dimagi/rd-toolkit/), you can customize this application in your current or future deployment of the CHT. 

Reviewer: When rereading this it seems as though it could use a bit more context to understand _why_ you'd want to use third-party applications.

```suggestion
Medic has worked with [FIND](https://www.finddx.org) to build a CHT reference application for COVID-19 point-of-care testing with Rapid Diagnostic Tests (RDT). Using the reference app as an example, CHT app developers can easily include the provisioning and capture of RDT in workflows. These workflows can include third-party applications, like [Dimagi's RD-Toolkit](https://github.com/dimagi/rd-toolkit/), that guide health workers through the use of the RDT.
```

You can find the code for this application in the [CHT Core repository on GitHub](https://github.com/medic/cht-core/tree/master/config/covid-19).

## Problem Being Addressed

The original call for proposals best describes why Medic created this app: 

> FIND is working towards supporting countries in implementing an effective test-trace-isolate response using digital tools. To this end, FIND is looking to partner with leading digital solution providers to accelerate the development and deployment of a set of minimum functionalities for the collection of COVID RDT data and supporting incorporation of these functionalities into existing digital tools for use in low- and middle-income country settings.

## Solution Overview

This reference app provides a base layer of functionality that you can easily customize to meet the needs of your health program.  Specifically, this application has examples of:

 * Provisioning COVID-19 RDT tests 
 * Capturing COVID-19 RDT results
 * Storing all RDT data, including pictures, in the CHT
 * Best practices for RDT workflows in the CHT
 

>  Your instance of CHT needs to be on v3.13+ which has features developed to support this reference application, including:

Reviewer: Is cht-android 0.10.0 implied, or should it be stated explicitly?

Reviewer: Having RD-Toolkit is a requirement for this particular app too, where can that be explained?

Author: Good questions both! Answered in 78e19ce.

<!-- 
  use raw HTML for now as hugo youtube short codes show 16:9 ration 
  box instead of mobile portrait ~9:16
-->
<iframe id="ytplayer" type="text/html" width="290" height="610" src="https://www.youtube.com/embed/FlhZ-KQIfdY" frameborder="0"></iframe>

#### Capture

This second video shows the right side of the workflow above to capture RDT results for a patient.  The CHW is shown viewing the "Capture Covid-19 RDT" task after the 15 minute timer from the RD-Toolkit app has completed. After selecting the task, the CHW is brought to the CHT form which has the session information for the RD-Toolkit already loaded.  The test results are recorded and then returned to the CHT from the RD-Toolkit. The last part of the video shows the completed capture report in the CHT, including the result and image of the RDT:

<!-- 
  use raw HTML for now as hugo youtube short codes show 16:9 ration 
  box instead of mobile portrait ~9:16
-->
<iframe id="ytplayer" type="text/html" width="290" height="610" src="https://www.youtube.com/embed/LC9gR3N8o80" frameborder="0"></iframe>

## Reports and Dashboards

> Like all applications written for the CHT, there are built in mechanism to retrieve raw and aggregate data to generate reports and dashboards. Here are some ways that the data can be accessed:

Reviewer: ```suggestion
Like all applications written for the CHT, there are built-in mechanisms to retrieve raw and aggregate data to generate reports and dashboards. Here are some ways that the data can be accessed:
```

 * **[In app targets]({{< ref "apps/features/targets" >}}):** Gives the CHW or their supervisor an aggregate view of any of the form fields. Since targets rely on the data on the device, if targets include data from other users then permissions must be set on the relevant forms so that the data can be replicated and synchronized accordingly.
 * **API Calls:** Given that all form submissions are captured in JSON and that we know the data model well, you can easily do API calls to a CHT server instance and use some custom code (node, python etc) to gather and show stats on a daily basis. You can export to either JSON or CSV. See API docs [for reports]({{< ref "apps/reference/api#get-apiv2exportreports" >}}) as well as [monitoring metadata]({{< ref "apps/reference/api#get-apiv2monitoring" >}}). 

>  * **[Postgres queries](https://github.com/medic/cht-couch2pg):** CHT ships with a utility to export all the data that the API has to a relational database, Postgres. You have all the raw data the API has, but can now use the power of joins and groupings to come up with totally customizable stats by day, week, month etc. Data can be synched daily or hourly from the CHT. 

Reviewer: Seems worthwhile to have the complete name once

```suggestion
 * **[PostgreSQL queries](https://github.com/medic/cht-couch2pg):** CHT ships with a utility to export all the data that the API has to a relational database, Postgres. You have all the raw data the API has, but can now use the power of joins and groupings to come up with totally customizable stats by day, week, month etc. Data can be synched daily or hourly from the CHT. 
```

>  * **Dashboards:** We have used both [Klipfolio](https://www.klipfolio.com/) and, more recently, [Superset](https://superset.incubator.apache.org/) to create more complex yet still user-friendly dashboards. This is particular useful for those who need to view the data but wouldn't otherwise be logging in to CHT apps. These dashboards generally access the relational data in the Postgres database as the back end.

Reviewer: ~We~

```suggestion
 * **Dashboards:** Medic has used both [Klipfolio](https://www.klipfolio.com/) and, more recently, [Superset](https://superset.incubator.apache.org/) to create more complex yet still user-friendly dashboards. This is particular useful for those who need to view the data but wouldn't otherwise be logging in to CHT apps. These dashboards generally access the relational data in the Postgres database as the back end.
```

]
```

The last two forms can be used to query [the reports API ]({{< ref "apps/reference/api#get-apiv2exportreports" >}}) to get RDT provions and captures.  To get all the capture reports into a file called `output.csv`, you would use this call:

```shell
curl "http://LOGIN:PASSWORD@HOSTNAME/api/v2/export/reports?filters[search]=&filters[forms][selected][0][code]=covid19_rdt_capture" > output.csv
```

{{% alert title="Note" %}}As all API calls need to be authenticated, be sure to use a login with admin permissions with this structure: `http://LOGIN:PASSWORD@HOSTNAME`.{{% /alert %}}

### Base64 image extraction

To extract the binary image from the ASCII value in a report, you can view it in the CHT or you can use a script to do so.  Here's an example of using Bash utilities (`sed`, `cut`, `tr` and `base64`) in Linux to loop over each line of the `output.csv` file from [the above call](#sample-api-call) to reports API. This results in numbered binary image files called `imageNUMBER.jpg` for each line in the export: 

```shell
i=1 n=0
while read -r line; do
  ((n >= i )) && echo $line | cut -d',' -f15 | tr -d '"' | base64 -d > image$n.jpg
  ((n++))
done < output.csv
```

> This code snippet is good just to validate data ad hoc, but more likely you'll be doing either an extract-transform-load (ETL) process into a 3rd party system, or just raw bulk import w/ no ETL and then on display show the base64 as an image, as we do in the CHT.

Reviewer: Suggestion for consistency

```suggestion
This code snippet is good just to validate data ad hoc, but more likely you'll be doing either an extract-transform-load (ETL) process into a third-party system, or just import the raw contents and then display the base64 as an image, as done in the CHT.
```

When retrieving [JSON](#capture-1), this value is found in `capture.android-app-outputs.rdt_session_bundle.rdt_session_result_bundle.rdt_session_result_extra_images.cropped` field.

{{% alert title="Note" %}}The reports API always outputs in CSV, but the COVID-19 application uses [Base64 encoding](https://en.wikipedia.org/wiki/Base64) to store the images as text as noted above.  These may misbehave when opened them in a spreadsheet application like LibreOffice ("maximum number of characters per cell exceeded") or Excel (silently clipped to 32k chars) as they're thousands, if not hundreds of thousands, of characters long.  Be sure to programmatically process these into image files as needed. {{% /alert %}}

> ## Customizing the reference app

Reviewer: Avoiding the use of "reference" here since it may be more of an "example" app

```suggestion
## Customizing the application
```

## Solution Overview

This reference app provides a base layer of functionality that you can easily customize to meet the needs of your health program.  Specifically, this application has examples of:

 * Provisioning COVID-19 RDT tests 
 * Capturing COVID-19 RDT results
 * Storing all RDT data, including pictures, in the CHT
 * Best practices for RDT workflows in the CHT
 
Your instance of CHT needs to be on v3.13+ which has features developed to support this reference application, including:
 
  * **Display Base64 Image** - show ASCII encoded images inline
  * **Android App Launcher** - integrate with third party apps like the RD-Toolkit
  * **Parse Timestamp to Date** -  convert epoch time stamps (`1628945040308`) to your desired format (`Sun Mar 19 13:25:08`).
  
For more information on these features, see the ["Related Content"](#related-content).

Additional requirements for this application beyond CHT 3.13, include [CHT Android 0.10.0](https://github.com/medic/cht-android) or later and Dimagi's [RD-Toolkit 0.9.8](https://github.com/dimagi/rd-toolkit/) or later. 

> While this application utilizes the RD-Toolkit, the integration features in the CHT Core and CHT Android are generic.  This means you could use a different RDT Android application if you prefer. Beyond the scope of RDTs, you may use any application that is supported.  To read more about this feature, see the [forms reference documentation]({{< ref "apps/reference/forms/app#android-app-launcher" >}}). 

Reviewer: Helpful new paragraph!

Personal preference to avoid using `utilize`, and have link text match the anchor more closely, along with a clarification:

```suggestion
While this application calls the RD-Toolkit, the integration features in the CHT Core and CHT Android are generic.  This means you could use a different RDT Android application if you prefer. Beyond the scope of RDTs, you could use this integration feature to launch any other Android app to perform an action and save the result in the CHT. To read more about this feature, see the [Android App Launcher section in the Forms reference documentation]({{< ref "apps/reference/forms/app#android-app-launcher" >}}). 
```
